iree/turbine/kernel/wave/scheduling/multi_buffering.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+from ..._support.tracing import CapturedTrace
+from ..._support.indexing import (
+    IndexSymbol,
+    xor,
+)
+from ...compiler.base import CodegenError
+from ...lang.global_symbols import SHARED_ADDRESS_SPACE
+from ...ops.wave_ops import (
+    get_custom,
+    Read,
+    Write,
+    CustomOp,
+    Reduction,
+)
+import iree.turbine.kernel.lang as tkl
+
+
+def multi_buffer(trace: CapturedTrace):
+    """Perform multi buffering for all supported shared memory locations"""
+
+    # Find all reductions
+    reductions = trace.walk(lambda node: isinstance(get_custom(node), Reduction))
+
+    # Get reduction dimension from first reduction
+    if not reductions or len(reductions) != 1:
+        raise CodegenError(
+            f"Unexpected number of reductions found in graph: {len(reductions)} vs 1"
+        )
+
+    reduction_axis = get_custom(reductions[0]).axis
+
+    # Find reads that index using the reduction dim and are from shared memory
+    reads = []
+    writes = []
+    for node in trace.get_subgraph(get_custom(reductions[0]).subgraph_name).nodes:
+        custom = get_custom(node)
+        if not hasattr(custom, "memory_type"):
+            continue
+        if (
+            reduction_axis in custom.indexing_dims
+            and custom.memory_type.address_space == SHARED_ADDRESS_SPACE
+        ):
+            if isinstance(custom, Read):
+                reads.append(custom)
+            elif isinstance(custom, Write):
+                writes.append(custom)
+
+    # Partition reads and writes by memory location
+    memory_to_reads = _partition_by_memory(reads)
+    memory_to_writes = _partition_by_memory(writes)
+
+    # Perform multi buffering for all collected memory locations
+    for memory_location in set(memory_to_reads.keys()) | set(memory_to_writes.keys()):
+        read_nodes = memory_to_reads.get(memory_location, [])
+        write_nodes = memory_to_writes.get(memory_location, [])
+
+        _multi_buffer_memory_location(
+            trace, memory_location, read_nodes, write_nodes, reduction_axis, 2
+        )
+
+
+def _multi_buffer_memory_location(
+    trace: CapturedTrace,
+    original_buffer: CustomOp,
+    read_nodes: list[Read],
+    write_nodes: list[Write],
+    axis: IndexSymbol,
+    buffer_count: int,
+):
+    """
+    Implements multi buffering for all reads and write of a shared memory buffer.
+    """
+    # For now we only support double buffering
+    if buffer_count != 2:
+        raise CodegenError(
+            "Current multi buffering implementation supports only buffer_count=2"
+        )
+
+    # double the memory in the non-reduction dimension
```

Comment:
why do you need this constraint? Instead of just looking at 1 - reduction_dim_index, could you partition the dims into reduction_axis and the rest?